src/Sarif/ZipArchiveArtifact.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+
+    public class ZipArchiveArtifact : IEnumeratedArtifact
+    {
+        private readonly ISet<string> binaryExtensions;
+        private readonly ZipArchive archive;
+        private ZipArchiveEntry entry;
+        private readonly Uri uri;
+        private string contents;
+        private byte[] bytes;
+
+        public ZipArchiveArtifact(ZipArchive archive, ZipArchiveEntry entry, ISet<string> binaryExtensions)
+        {
+            this.entry = entry;
+            this.archive = archive;
+            this.binaryExtensions = binaryExtensions;
+            this.uri = new Uri(entry.FullName, UriKind.RelativeOrAbsolute);
+        }
+
+        public Uri Uri => this.uri;
+
+        public Stream Stream
+        {
+            get
+            {
+                lock (this.archive)
```

Comment:
Maybe strengthen error checking in the constructor to preclude the null condition and elide the lock?

Reply:
Bonus: if we do need locking, it would be nice to be locking the entry as opposed to the archive.  (Assuming you don't need a whole archive locked to safely open an entry and such.)

Reply:
You have to lock on the archive! Too bad. We tested the other case. There must be (makes sense) state maintained at the archive level about decompression/other entry state. I'm not sure why we permit a null entry. Will try to rip that out.